src/testing/testing.go

```diff
@@ -204,35 +208,70 @@ func (c *common) Parallel() {
 	// Unimplemented.
 }
 
-// Run runs a subtest of f t called name. It waits until the subtest is finished
-// and returns whether the subtest succeeded.
-func (t *T) Run(name string, f func(t *T)) bool {
-	// Create a subtest.
-	sub := T{
-		common: common{
-			name:   t.name + "/" + name,
-			indent: t.indent + "    ",
-		},
+// Cleanup registers a function to be called when the test (or subtest) and all its
+// subtests complete. Cleanup functions will be called in last added,
+// first called order.
+func (c *common) Cleanup(f func()) {
+	c.cleanups = append(c.cleanups, f)
+}
+
+// runCleanup is called at the end of the test.
+func (c *common) runCleanup() {
+	for {
+		var cleanup func()
+		if len(c.cleanups) > 0 {
+			last := len(c.cleanups) - 1
+			cleanup = c.cleanups[last]
+			c.cleanups = c.cleanups[:last]
+		}
+		if cleanup == nil {
+			return
+		}
+		cleanup()
 	}
+}
+
+func tRunner(t *T, fn func(t *T)) {
+	defer func() {
+		t.runCleanup()
+	}()
 
 	// Run the test.
 	if flagVerbose {
-		fmt.Fprintf(&t.output, "=== RUN   %s\n", sub.name)
-
+		fmt.Fprintf(t.w, "=== RUN   %s\n", t.name)
 	}
-	f(&sub)
+
+	fn(t)
 
 	// Process the result (pass or fail).
-	if sub.failed {
-		t.failed = true
-		fmt.Fprintf(&t.output, sub.indent+"--- FAIL: %s\n", sub.name)
-		t.output.Write(sub.output.Bytes())
+	if t.failed {
+		if t.parent != nil {
+			t.parent.failed = true
+		}
+		fmt.Fprintf(t.w, t.indent+"--- FAIL: %s\n", t.name)
+		t.w.Write(t.output.Bytes())
 	} else {
 		if flagVerbose {
-			fmt.Fprintf(&t.output, sub.indent+"--- PASS: %s\n", sub.name)
-			t.output.Write(sub.output.Bytes())
+			fmt.Fprintf(t.w, t.indent+"--- PASS: %s\n", t.name)
+			t.w.Write(t.output.Bytes())
 		}
 	}
+}
+
+// Run runs a subtest of f t called name. It waits until the subtest is finished
```

Comment:
Seems like this comment is missing a word or two?

Reply:
Repaired word salad.